src/ring_theory/unique_factorization_domain.lean

```diff
@@ -430,6 +448,218 @@ begin
+    revert hg,
+    change ((factors (associates.mk a)) ⊓ (factors (associates.mk b))).prod ≠ 1,
+    rw ← multiset.prod_zero, apply ne.elim, apply mt (@eq_of_prod_eq_prod _ _ _inst_2 _ 0),
+    rw factors_mk _ ha0, rw factors_mk _ hb0,
+    rw ← with_top.coe_inf, apply (mt with_top.coe_eq_coe.mp),
+    apply mt multiset.eq_zero_iff_forall_not_mem.mp, apply not_forall.mpr,
+    use ⟨associates.mk p, hp2⟩, apply not_not.mpr,
+    rw multiset.inf_eq_inter,
+    apply multiset.mem_inter.mpr
+      (and.intro (mem_factors'_of_dvd ha0 hp1 ha) (mem_factors'_of_dvd hb0 hp1 hb)) },
+  { intro hc,
+    by_contradiction hg,
+    obtain ⟨p, ⟨hp, ⟨hpa, hpb⟩⟩⟩ := exists_prime_dvd_of_not_inf_one ha0 hb0 hg,
+    apply absurd hp (hc hpa hpb) }
+end
+
+theorem prime_pow_factors (p : associates α) (h₂ : irreducible p)
+ (k : ℕ) : factors (p ^ k) = some (multiset.repeat ⟨p, h₂⟩ k) :=
+begin
+  apply eq_of_prod_eq_prod, rw associates.factors_prod,
+  unfold factor_set.prod, rw multiset.map_repeat,
+  rw multiset.prod_repeat, refl
+end
+
+theorem dvd_prime_pow_iff_count_factors (m : associates α) (h₁ : m ≠ 0) (p : associates α)
+  (h₂ : irreducible p) (k : ℕ) : p ^ k ≤ m ↔ k ≤ count ⟨p, h₂⟩ m.factors :=
+begin
+  obtain ⟨a, nz, h⟩ := associates.exists_non_zero_rep h₁,
+  erw [← h, (factors_mk a nz), ← with_top.some_eq_coe, multiset.le_count_iff_repeat_le],
+  split,
+  { intro h3,
+    apply with_top.coe_le_coe.1,
+    rw [← with_top.some_eq_coe, ← (prime_pow_factors p h₂), ← (factors_mk a nz)],
+    exact factors_le.2 h3 },
+  { intro h3,
+    apply factors_le.1,
+    erw [factors_mk a nz, prime_pow_factors p h₂],
+    exact with_top.coe_le_coe.2 h3 }
+end
+
+theorem count_product {a : associates α} (ha : a ≠ 0) {b : associates α} (hb : b ≠ 0)
+  {p : { a : associates α // irreducible a }}
+  : count p a.factors + count p b.factors = count p (factors (a * b)) :=
+begin
+  obtain ⟨a0, nza, ha'⟩ := exists_non_zero_rep ha,
+  obtain ⟨b0, nzb, hb'⟩ := exists_non_zero_rep hb,
+  erw [factors_mul, ← ha', ← hb', factors_mk a0 nza, factors_mk b0 nzb],
+  rw [← factor_set.coe_add, ← with_top.some_eq_coe, ← with_top.some_eq_coe, ← with_top.some_eq_coe],
+  unfold count, rw multiset.count_add
+end
+
+theorem count_of_coprime {a : associates α} (ha : a ≠ 0) {b : associates α} (hb : b ≠ 0)
+  (hab : ∀ (d), d ∣ a → d ∣ b → ¬ prime d) (p : { a : associates α // irreducible a })
+  : count p a.factors = 0 ∨ count p b.factors = 0 :=
+begin
+  cases nat.eq_zero_or_pos (count p a.factors) with hz h2, { exact or.inl hz },
+  have h3 : p.val ∣ a,
+  { rw ← (pow_one p.val), apply (dvd_prime_pow_iff_count_factors a ha p.val p.2 1).2,
+    simpa only [subtype.eta] },
+  apply or.intro_right,
+  cases nat.eq_zero_or_pos (count p b.factors) with hz h4, { exact hz },
+  have h5 : p.val ∣ b,
+  { rw ← (pow_one p.val),
+    apply (dvd_prime_pow_iff_count_factors b hb p.val p.2 1).2, simpa only [subtype.eta] },
+  apply (absurd (associates_irreducible_iff_prime.mp p.2) (hab p h3 h5))
+end
```

Comment:
```suggestion
begin
  rw [or_iff_not_imp_left, ← ne.def, ← nat.pos_iff_ne_zero],
  intro hca,
  contrapose! hab with hcb,
  rw ← nat.pos_iff_ne_zero at hcb,
  have hpa : p.val ∣ a,
  { rw [← pow_one p.val],
    apply (prime_pow_dvd_iff_le ha p.2).2,
    simpa only [subtype.eta] },
  have hpb : p.val ∣ b,
  { rw [← pow_one p.val],
    apply (prime_pow_dvd_iff_le hb p.2).2,
    simpa only [subtype.eta] },
  exact ⟨p, hpa, hpb, (associates_irreducible_iff_prime.mp p.2)⟩,
end
```

Maybe `0 < count p a.factors -> p | a` and `0 ≠ count p a.factors -> p | a` should be a separate lemma (we just need `dvd_of_mem_factors`, not even `prime_pow_dvd_iff_le`), then we can remove the duplication in the proof.